### PR TITLE
jerryx: add jerryx_arg_array

### PR DESCRIPTION
**Pull request by jiangzidong**
Related issue: #2046

JerryScript-DCO-1.0-Signed-off-by: Zidong Jiang zidong.jiang@intel.com

### CODE REVIEW

**Comment by martijnthe** [comment]
Sweet, thanks @jiangzidong!

@akosthekiss mind reviewing this?

File: tests/unit-ext/test-ext-arg.c [hunk 1]
```diff
@@ -55,6 +55,10 @@ const char *test_source = (
                            "                    -1000, 4294967297, 65536, -2200000000, 4294967297, -2147483647);"
                            "test_validator_int2(-1.5, -1.5, -1.5, 1.5, 1.5, 1.5, Infinity, -Infinity, 300.5, 300.5);"
                            "test_validator_int3(NaN);"
+                           "var arr = [1,2];"
```
Comment (akosthekiss) [inline tests/unit-ext/test-ext-arg.c]: (style) add a space after the comma

File: tests/unit-ext/test-ext-arg.c [hunk 2]
```diff
@@ -489,6 +493,77 @@ test_validator_int3_handler (const jerry_value_t func_obj_val __attribute__((unu
...
+
+  arr_info.c_arg_p = item_mapping;
+  arr_info.c_arg_cnt = 3;
+
+  jerryx_arg_t mapping[] =
+  {
+    jerryx_arg_array (&arr_info, JERRYX_ARG_OPTIONAL),
+  };
+
+  jerry_value_t is_ok = jerryx_arg_transform_args (args_p, 1, mapping, 1);
```
Comment (akosthekiss) [inline tests/unit-ext/test-ext-arg.c]: Is `(args_p, 1,` OK? Shouldn't this be `(args_p, args_cnt,`. I see that we have a similar construct in `test_validator_prop2_handler`, but actually that does not seem right to me either.

File: jerry-ext/include/jerryscript-ext/arg.h [hunk 3]
```diff
@@ -54,6 +54,15 @@ typedef struct
 } jerryx_arg_object_props_t;
 
 /**
+ * The structure used in jerryx_arg_array
+ */
+typedef struct
+{
+  const jerryx_arg_t *c_arg_p; /**< points to the array of transformation steps */
+  jerry_length_t c_arg_cnt; /**< the count of the `c_arg_p` array */
+} jerryx_arg_array_props_t;
```
[the remark this answers is not in the document]
Comment (jiangzidong) [reply jerry-ext/include/jerryscript-ext/arg.h]: I will choose the `jerryx_arg_array_items_t`, thanks!

File: tests/unit-ext/test-ext-arg.c (hunk 2 above)
Comment (jiangzidong) [reply tests/unit-ext/test-ext-arg.c]: I will change them all and avoid the magic number, thanks!

Comment (jiangzidong) [reply tests/unit-ext/test-ext-arg.c]: I will add a macro to get the item count of the mapping
```
#define ARRAY_SIZE(array) ((unsigned long)(sizeof(array) / sizeof((array)[0])))
```

**Comment by jiangzidong** [comment]
updated. @akosthekiss

File: docs/09.EXT-REFERENCE-ARG.md [hunk 4]
```diff
@@ -580,6 +630,97 @@ my_external_handler (const jerry_value_t function_obj,
...
+Create a validation/transformation step (`jerryx_arg_t`) that expects to
+consume one `array` JS argument and call `jerryx_arg_transform_array_items` inside
+to transform its items to native arguments.
+User should prepare the `jerryx_arg_array_items_t` instance, and pass it to this function.
+
+**Prototype**
+
+```c
+static inline jerryx_arg_t
+jerryx_arg_array (const jerryx_arg_array_items_t *array_props_p, jerryx_arg_optional_t opt_flag);
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: still props

File: jerry-ext/include/jerryscript-ext/arg.h [hunk 5]
```diff
@@ -162,6 +174,8 @@ static inline jerryx_arg_t
 jerryx_arg_custom (void *dest, uintptr_t extra_info, jerryx_arg_transform_func_t func);
 static inline jerryx_arg_t
 jerryx_arg_object_properties (const jerryx_arg_object_props_t *object_props_p, jerryx_arg_optional_t opt_flag);
+static inline jerryx_arg_t
+jerryx_arg_array (const jerryx_arg_array_items_t *array_props_p, jerryx_arg_optional_t opt_flag);
```
Comment (akosthekiss) [inline jerry-ext/include/jerryscript-ext/arg.h]: still props

File: tests/unit-ext/test-common.h [hunk 6]
```diff
@@ -18,6 +18,8 @@
 
 #include "jerryscript-port.h"
 
+#define ARRAY_SIZE(array) ((unsigned long)(sizeof(array) / sizeof((array)[0])))
```
Comment (akosthekiss) [inline tests/unit-ext/test-common.h]: (style) add space after type cast, add space after sizeofs

**Pull request merged**